runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -0,0 +1,229 @@
+     */
+
+    public static List<Long> appendCheckpoint(CorfuRuntime rt, UUID streamID, String author, SMRMap map,
+                                              Consumer<CheckpointWriter> setupWriter) {
+        List<Long> addrs = new ArrayList<>();
+        CheckpointWriter cpw = new CheckpointWriter(rt, streamID, author, map);
+
+        setupWriter.accept(cpw);
+        addrs.add(cpw.startCheckpoint());
+        addrs.addAll(cpw.appendObjectState());
+        addrs.add(cpw.finishCheckpoint());
+        return addrs;
+    }
+
+    /** Append a checkpoint START record to this object's stream.
+     *  TX side-effect: We call TXBegin().
+     *
+     * @return Global log address of the START record.
+     */
+    public long startCheckpoint() {
+        rt.getObjectsView().TXBegin();
+        startTime = LocalDateTime.now();
+        AbstractTransactionalContext context = TransactionalContext.getCurrentContext();
+        long txBeginGlobalAddress = context.getSnapshotTimestamp();
+
+        this.mdKV.put(CheckpointEntry.START_TIME, startTime.toString());
+        this.mdKV.put(CheckpointEntry.START_LOG_ADDRESS, Long.toString(txBeginGlobalAddress));
+
+        ImmutableMap<String,String> mdKV = ImmutableMap.copyOf(this.mdKV);
+        CheckpointEntry cp = new CheckpointEntry(CheckpointEntry.CheckpointEntryType.START,
+                author, checkpointID, mdKV, null);
+        startAddress = sv.append(cp, null, null);
+        postAppendFunc.accept(cp, startAddress);
+        return startAddress;
+    }
+
+    /** Append zero or more CONTINUATION records to this
+     *  object's stream.  Each will contain a fraction of
+     *  the state of the object that we're checkpointing
+     *  (up to batchSize items at a time).
+     *
+     *  The Iterators class appears to preserve the laziness
+     *  of Stream processing; we don't wish to use more
+     *  memory than strictly necessary to generate the
+     *  checkpoint.  NOTE: It would be even more useful if
+     *  the map had a lazy iterator: the eagerness of
+     *  map.keySet().stream() is not ideal, but at least
+     *  it should be much smaller than the entire map.
+     *
+     *  NOTE: The postAppendFunc lambda is executed in the
+     *  current thread context, i.e., inside of a Corfu
+     *  transaction, and that transaction will be *aborted*
+     *  at the end of this function.  Any Corfu data
+     *  modifying ops will be undone by the TXAbort().
+     *
+     *  TX side-effect: We call TXAbort().
+     *
+     * @return Stream of global log addresses of the
+     * CONTINUATION records written.
+     */
+    public List<Long> appendObjectState() {
+        ImmutableMap<String,String> mdKV = ImmutableMap.copyOf(this.mdKV);
+        List<Long> continuationAddresses = new ArrayList<>();
+
+        try {
+            Iterators.partition(map.keySet().stream()
+                    .map(k -> {
+                        return new SMREntry("put",
```

Comment:
Dahlia asked the same question, and my reply was "we need either Nx1 or 1xN of key-value pairs".  The checkpoint data is only applied once at most, is it really worth optimizing here?